src/ParallelAlgorithms/Amr/Criteria/Loehner.hpp

```diff
@@ -0,0 +1,269 @@
+ */
+template <size_t Dim>
+double loehner_smoothness_indicator(
+    gsl::not_null<DataVector*> first_deriv_buffer,
+    gsl::not_null<DataVector*> second_deriv_buffer,
+    const DataVector& tensor_component, const Mesh<Dim>& mesh,
+    size_t dimension);
+template <size_t Dim>
+std::array<double, Dim> loehner_smoothness_indicator(
+    const DataVector& tensor_component, const Mesh<Dim>& mesh);
+/// @}
+
+namespace Loehner_detail {
+template <size_t Dim>
+void max_over_components(
+    gsl::not_null<std::array<Flag, Dim>*> result,
+    gsl::not_null<std::array<DataVector, 2>*> deriv_buffers,
+    const DataVector& tensor_component, const Mesh<Dim>& mesh,
+    double relative_tolerance, double absolute_tolerance,
+    double coarsening_factor);
+}
+
+/*!
+ * \brief h-refine the grid based on a smoothness indicator
+ *
+ * The smoothness indicator used here is based on the magnitude of second
+ * derivatives. See `amr::Criteria::loehner_smoothness_indicator` for details
+ * and caveats.
+ *
+ * \see amr::Criteria::loehner_smoothness_indicator
+ */
+template <size_t Dim, typename TensorTags>
+class Loehner : public Criterion {
+ public:
+  struct VariablesToMonitor {
+    using type = std::vector<std::string>;
+    static constexpr Options::String help = {
+        "The tensors to monitor for h-refinement."};
+    static size_t lower_bound_on_size() { return 1; }
+  };
+  struct RelativeTolerance {
+    using type = double;
+    static constexpr Options::String help = {
+        "If any tensor component has a second derivative magnitude above this "
+        "value times the max of the absolute tensor component over the "
+        "element, the element will be h-refined in that direction. "
+        "Set to 0 to disable."};
+    static double lower_bound() { return 0.; }
+  };
+  struct AbsoluteTolerance {
+    using type = double;
+    static constexpr Options::String help = {
+        "If any tensor component has a second derivative magnitude above this "
+        "value, the element will be h-refined in that direction. "
+        "Set to 0 to disable."};
+    static double lower_bound() { return 0.; }
+  };
+  struct CoarseningFactor {
+    using type = double;
+    static constexpr Options::String help = {
+        "Factor applied to both relative and absolute tolerance to trigger "
+        "h-coarsening. Set to 0 to disable h-coarsening altogether. "
+        "Set closer to 1 to trigger h-coarsening more aggressively. "
+        "Values too close to 1 risk that coarsened elements will immediately "
+        "trigger h-refinement again. A reasonable value is 1/3."};
+    static double lower_bound() { return 0.; }
+    static double upper_bound() { return 1.; }
+  };
+
+  using options = tmpl::list<VariablesToMonitor, RelativeTolerance,
+                             AbsoluteTolerance, CoarseningFactor>;
+
+  static constexpr Options::String help = {
+      "Refine the grid towards the target truncation error"};
```

Comment:
maybe "towards resolving an estimated error in the second derivative."?

Reply:
Oh good catch, this was still copied text